libavcodec/jpeg2000dec.c

```diff
@@ -1067,100 +1072,339 @@ static int jpeg2000_decode_packet(Jpeg2000DecoderContext *s, Jpeg2000Tile *tile,
-                    int href_passes =  (cblk->npasses + newpasses - 1) % 3;
-                    int eb = av_log2(newpasses - href_passes);
-                    int extra_bit = newpasses > 2 ? 1 : 0;
-                    if ((ret = get_bits(s, llen + eb + 3)) < 0)
-                        return ret;
-                    cblk->pass_lengths[0] = ret;
-                    if ((ret = get_bits(s, llen + 3 + extra_bit)) < 0)
-                        return ret;
-                    cblk->pass_lengths[1] = ret;
-                    ret = cblk->pass_lengths[0] + cblk->pass_lengths[1];
+                // Update cblk->npasses and write length information
+                cblk->npasses = (uint8_t) (cblk->npasses + segment_passes);
+                segment_length[nb_segments++] = segment_bytes;
+
+                if ((cblk->modes & JPEG2000_CTSY_HTJ2K_F) && cblk->ht_plhd == HT_PLHD_OFF) {
+                    newpasses -= (uint8_t) segment_passes;
+                    primary_passes = (uint8_t) (segment_passes + cblk->fast_skip_passes);
+                    cblk->fast_skip_passes = 0;
+                    primary_bytes = segment_bytes;
+                    secondary_passes = 0;
+                    secondary_bytes = 0;
+                    empty_set = 0;
+                    if (next_segment_passes == 2 && segment_bytes == 0)
+                        empty_set = 1;
+                    while (newpasses > 0) {
+                        if (newpasses > 1)
+                            segment_passes = next_segment_passes;
+                        else
+                            segment_passes = 1;
+                        next_segment_passes = (uint8_t) (3 - next_segment_passes);
+                        bits_to_read =
+                                (uint8_t) (cblk->lblock + (unsigned int) (segment_passes) - 1);
+                        segment_bytes = get_bits(s, bits_to_read);
+                        newpasses -= (uint8_t) (segment_passes);
+                        if (next_segment_passes == 2) {
+                            // This is a FAST Cleanup pass
+                            if(segment_passes != 1)
+                                av_log(s->avctx, AV_LOG_WARNING, "Number of HT cleanup pass shall be one.\n");
+                            if (segment_bytes != 0) {
+                                // This will have to be the new primary
+                                if (segment_bytes < 2)
+                                    av_log(s->avctx, AV_LOG_WARNING, "Length information for a HT-codeblock is invalid\n");
+                                // Write length information for HT CleanUp segment
+                                cblk->pass_lengths[0] = segment_bytes; 
+                                fast_skip_bytes += primary_bytes + secondary_bytes;
+                                primary_passes++;
+                                primary_passes = (uint8_t) (primary_passes + secondary_passes);
```

Comment:
See comments on bits_to_read

Reply:
This section of code was for parsing multiple HT sets. Since multiple HT sets is not supported for HT block decoding in FFmpeg, I removed the related lines. This removal does not trigger any regression.